File: src/Algorithms/Lists/Three_Computational_Problems/best_sum.py
def best_sum(target_sum, numbers):
    if target_sum < 0:
        return None
    if target_sum == 0:
        return []

    shortest_combination = None

    for num in numbers:
        remainder = target_sum - num
        combination = best_sum(remainder, numbers)

        if combination is not None:
            combination.append(num)

            if shortest_combination is None or len(shortest_combination) > len(combination):
                shortest_combination = combination

    return shortest_combination

File: src/Algorithms/Lists/Three_Computational_Problems/best-sum.ts
// Brute Force:
// O(n^m * m) time
// O(m^2) space
// Where m = target sum, n = array length

// Memoized:
// O(m^2 * n) time
// O(m^2) space
const bestSumMemo = (
  targetSum: number,
  numbers: number[],
  memo: Record<number, number[] | null> = {}
): number[] | null => {
  if (targetSum in memo) return memo[targetSum];
  if (targetSum === 0) return [];
  if (targetSum < 0) return null;

  let shortestCombination = null;

  for (let num of numbers) {
    const remainder = targetSum - num;
    const remainderCombination = bestSumMemo(remainder, numbers, memo);
    if (remainderCombination !== null) {
      const combination = [...remainderCombination, num];

      if (
        !shortestCombination ||
        shortestCombination.length > combination.length
      ) {
        shortestCombination = combination;
      }
    }
  }

  memo[targetSum] = shortestCombination;
  return shortestCombination;
};

export const solutions = {
  bestSumMemo,
};
